emulator.c

```diff
@@ -668,3 +673,36 @@ print_flags(uint8_t flags)
          (flags & FLAG_S) == FLAG_S, (flags & FLAG_P) == FLAG_P,
          (flags & FLAG_CY) == FLAG_CY, (flags & FLAG_AC) == FLAG_AC);
 }
+
+// INTERRUPTS
+int
+handle_interrupt(i8080 *cpu, uint8_t rst_instruction)
+{
+  if (rst_instruction > RST_RANGE)
+    {
+      fprintf(stderr, "Invalid restart instruction %u", rst_instruction);
+      return -1;
+    }
+
+  if (cpu->interrupt_enabled == false)
+    {
+      return 0;
+    }
+
+  // get address for interrupt subroutine
+  uint16_t subroutine_address = BYTE * rst_instruction;
+
+  // push program counter to stack
+  cpu_write_mem(cpu, cpu->sp - 1,
+                (uint8_t)((cpu->pc & UPPER_8_BIT_MASK) >> BYTE));
+  cpu_write_mem(cpu, cpu->sp - 2, (uint8_t)(cpu->pc & LOWER_8_BIT_MASK));
+  cpu->sp -= 2;
+
+  // set program counter to start of the interrupt subroutine
+  cpu->pc = subroutine_address;
```

Comment:
So the RST subroutine isn't an opcode, but the instructions after the subroutine are all opcodes that we implemented?And they will be able to correctly handle the pushes that you have pushed to the stack?

Reply:
The RST subroutine is a set of instructions written into the Space Invaders ROM - all of which have opcodes we've implemented. All this does is set the PC to the starting address of that subroutine written by the space invaders developers.